resourcemanager/resourceids/errors.go

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package resourceids
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+)
+
+var _ error = SegmentNotSpecifiedError{}
+
+type SegmentNotSpecifiedError struct {
+	parseResult    ParseResult
+	resourceId     ResourceId
+	resourceIdName string
+	segmentName    string
+}
+
+// NewSegmentNotSpecifiedError returns a SegmentNotSpecifiedError for the provided Resource ID, segment and parseResult combination
+func NewSegmentNotSpecifiedError(id ResourceId, segmentName string, parseResult ParseResult) SegmentNotSpecifiedError {
+	// Resource ID types must be in the format {Name}Id
+	resourceIdName := strings.TrimSuffix(reflect.ValueOf(id).Type().Name(), "Id")
+	return SegmentNotSpecifiedError{
+		resourceIdName: resourceIdName,
+		resourceId:     id,
+		segmentName:    segmentName,
+		parseResult:    parseResult,
+	}
+}
+
+// Error returns a detailed error message highlighting the issues found when parsing this Resource ID Segment.
+func (e SegmentNotSpecifiedError) Error() string {
+	expectedId := e.buildExpectedResourceId()
+	description, err := e.descriptionForSegment(e.segmentName)
+	if err != nil {
+		// this isn't user recoverable, and will be caught at unit/acceptance test time, so `panic` is reasonable here
+		return fmt.Sprintf("internal-error: building description for segment: %+v", err)
+	}
+
+	return fmt.Sprintf(`parsing the %[1]s ID: the segment %[2]q was not found
+
+Expected a %[1]s ID that matched:
+
+> %[3]s
+
+However this value was provided:
+
+> %[4]s
+
+However the parsed Resource ID was missing a value for the %[2]q segment
```

Comment:
This reads weird with the `However` up above. I think we can just drop it
```suggestion
The parsed Resource ID was missing a value for the %[2]q segment
```